### PR TITLE
csi: add topology provisioning support

### LINKED ISSUE #10935: Support for Topolgy based provisoning
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
We are adding support for Replica-1 Non-Resilient pools for ocs/odf.

But we currently don't have support from rook side, for it.
So we have to manually patch things like rbac cluster roles & deployments to get the thing working.

We are using this.

`#Patch rbacs
CLUSTER_ROLE1=$(kubectl get clusterrolebinding  -o custom-columns='NAME:roleRef.name,SERVICE_ACCOUNTS:subjects[?(@.kind=="ServiceAccount")].name' | grep "rook-csi-rbd-provisioner-sa" |awk '{print $1}')
	oc patch clusterrole "$CLUSTER_ROLE1" --type=json -p='[{"op": "add", "path": "/rules/-","value": {"apiGroups":[""],"resources":["nodes"],"verbs": ["get", "list", "watch"]}}]'
	oc patch clusterrole "$CLUSTER_ROLE1" --type=json -p='[{"op": "add", "path": "/rules/-","value": {"apiGroups":["storage.k8s.io"],"resources":["csinodes"],"verbs": ["get", "list", "watch"]}}]'

CLUSTER_ROLE2=$(kubectl get clusterrolebinding  -o custom-columns='NAME:roleRef.name,SERVICE_ACCOUNTS:subjects[?(@.kind=="ServiceAccount")].name' | grep "rook-csi-rbd-plugin-sa" |awk '{print $1}')
	oc patch clusterrole "$CLUSTER_ROLE2" --type=json -p='[{"op": "add", "path": "/rules/-","value": {"apiGroups":[""],"resources":["nodes"],"verbs": ["get"]}}]'

#Patch csi-rbdplugin-provisioner deployment
INDEX1=$(oc get deployment csi-rbdplugin-provisioner -o json  | jq ' .spec.template.spec.containers | map(.name == "csi-provisioner") | index(true)')
	oc patch deployment csi-rbdplugin-provisioner --type=json -p='[{"op": "add", "path": '/spec/template/spec/containers/"$INDEX1"/args/-', "value": "--feature-gates=Topology=true" }]'

#Patch csi-rbdplugin daemonset
	INDEX2=$(oc get daemonset csi-rbdplugin -o json  | jq ' .spec.template.spec.containers | map(.name == "csi-rbdplugin") | index(true)')
	oc patch daemonset csi-rbdplugin --type=json -p='[{"op": "add", "path": '/spec/template/spec/containers/"$INDEX2"/args/-', "value": "--domainlabels=topology.kubernetes.io/region,topology.kubernetes.io/zone" }]'`

[Here](https://github.com/ceph/ceph-csi/pull/816) is the PR where Topology aware pool support was added to ceph-csi
[Here](https://gist.github.com/ShyamsundarR/8ff154555f137e71e5c0a80e224c61a2) is a doc to better understand this
[Here](https://github.com/ceph/ceph-csi/issues/440) is the issue on ceph-csi side tracking the progress of the feature

### PR DESCRIPTION
**Pull request by Rakshith-R**
This commit adds topology provisioning
support. This makes modification to rbac,
csi deployment and daemonset.

Signed-off-by: Rakshith R <rar@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #10935

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/csi/nfs/rbac.yaml [hunk 1]
```diff
@@ -16,6 +16,9 @@ rules:
   - apiGroups: [""]
     resources: ["nodes"]
     verbs: ["get"]
+  - apiGroups: [""]
```
Comment (Madhu-1) [inline deploy/examples/csi/nfs/rbac.yaml:19]: why nfs rbac change?

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml [hunk 2]
```diff
@@ -93,6 +93,9 @@ spec:
             - "--leader-election-renew-deadline=107s"
             - "--leader-election-retry-period=26s"
             - "--extra-create-metadata=true"
+            {{- if .EnableCSITopology }}
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml:96]: cephfs doesnt support it.

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml [hunk 3]
```diff
@@ -67,6 +67,9 @@ spec:
             - "--forcecephkernelclient={{ .ForceCephFSKernelClient }}"
             - "--metricspath=/metrics"
             - "--enablegrpcmetrics={{ .EnableCSIGRPCMetrics }}"
+            {{- if .EnableCSITopology }}
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:70]: cephfs doesnt support it.

File: deploy/examples/common.yaml [hunk 4]
```diff
@@ -91,6 +97,9 @@ rules:
   - apiGroups: [""]
     resources: ["serviceaccounts/token"]
     verbs: ["create"]
+  - apiGroups: [""]
```
Comment (Madhu-1) [inline deploy/examples/common.yaml:100]: do we need this change?

File: deploy/charts/rook-ceph/templates/configmap.yaml [hunk 5]
```diff
@@ -82,6 +82,12 @@ data:
   ROOK_CSIADDONS_IMAGE: {{ .Values.csi.csiAddons.image | quote }}
 {{- end }}
 {{- end }}
+{{- if .Values.csi.topology }}
+  CSI_ENABLE_TOPOLOGY: {{ .Values.csi.topology.enabled | quote }}
```
Comment (Madhu-1) [inline deploy/charts/rook-ceph/templates/configmap.yaml:86]: for now keep this change specific to RBD not a common for both cephfs and rbd

File: deploy/examples/csi/nfs/rbac.yaml (hunk 1 above)
Comment (Rakshith-R) [reply deploy/examples/csi/nfs/rbac.yaml]: by mistake, removed it.

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml (hunk 2 above)
Comment (Rakshith-R) [reply pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin-provisioner-dep.yaml]: removed it.

File: pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml (hunk 3 above)
Comment (Rakshith-R) [reply pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml]: removed it.

File: deploy/examples/common.yaml (hunk 4 above)
Comment (Rakshith-R) [reply deploy/examples/common.yaml:94]: yes https://github.com/rook/rook/issues/10935

File: deploy/charts/rook-ceph/templates/configmap.yaml (hunk 5 above)
Comment (Rakshith-R) [reply deploy/charts/rook-ceph/templates/configmap.yaml:86]: all this now only applies to RBD

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml [hunk 6]
```diff
@@ -36,6 +36,9 @@ spec:
             - "--leader-election-retry-period=26s"
             - "--default-fstype=ext4"
             - "--extra-create-metadata=true"
+            {{- if .EnableCSITopology }}
+            - --feature-gates=Topology=true"
```
Comment (malayparida2000) [inline pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml:40]: Need a double quote here 

`- "--feature-gates=Topology=true"`

File: deploy/charts/rook-ceph/values.yaml [hunk 7]
```diff
@@ -384,6 +384,16 @@ csi:
...
+  # Enable topology based provisioning.
+  topology:
+    enabled: false
+    # domainLabels define which node labels to use as domains
+    # for CSI nodeplugins to advertise their domains
+    # NOTE: the value here serves as an example and needs to be
+    # updated with node labels that define domains of interest
+    # domainLabels:
+    # - topology.kubernetes.io/region
+    # - topology.kubernetes.io/zone
```
[the remark this answers is not in the document]
Comment (Rakshith-R) [reply deploy/charts/rook-ceph/values.yaml:397]: added hostname and rook/ rack ,

We can add doc link with malay's pr for #10944

File: pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml (hunk 6 above)
Comment (Rakshith-R) [reply pkg/operator/ceph/csi/template/rbd/csi-rbdplugin-provisioner-dep.yaml]: thanks, done

File: deploy/charts/rook-ceph/values.yaml [hunk 8]
```diff
@@ -386,6 +386,17 @@ csi:
...
+  # Enable topology based provisioning.
+  topology:
+    enabled: false
+    # domainLabels define which node labels to use as domains
+    # for CSI nodeplugins to advertise their domains
+    # NOTE: the value here serves as an example and needs to be
+    # updated with node labels that define domains of interest
+    # domainLabels:
+    # - topology.kubernetes.io/zone
+    # - topology.kubernetes.io/hostname
```
Comment (malayparida2000) [inline deploy/charts/rook-ceph/values.yaml:398]: This should be
` - kubernetes.io/hostname`

File: deploy/charts/rook-ceph/values.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (malayparida2000) [reply deploy/charts/rook-ceph/values.yaml:397]: Let's keep

```
domainLabels:
- kubernetes.io/hostname
- topology.kubernetes.io/region
- topology.kubernetes.io/zone
- topology.rook.io/rack

# Link to full rook topology labels https://rook.io/docs/rook/latest/CRDs/Cluster/ceph-cluster-crd/#other-status
```

File: deploy/examples/operator-openshift.yaml [hunk 9]
```diff
@@ -524,6 +524,14 @@ data:
...
   # The GCSI RPC timeout value (in seconds). It should be >= 120. If this variable is not set or is an invalid value, it's default to 150.
   CSI_GRPC_TIMEOUT_SECONDS: "150"
+
+  # Enable topology based provisioning.
+  CSI_ENABLE_TOPOLOGY: "false"
+  # Domain labels define which node labels to use as domains
+  # for CSI nodeplugins to advertise their domains
+  # NOTE: the value here serves as an example and needs to be
+  # updated with node labels that define domains of interest
+  # CSI_TOPOLOGY_DOMAIN_LABELS: "topology.kubernetes.io/zone,topology.kubernetes.io/hostname,topology.rook.io/rack"
```
Comment (malayparida2000) [inline deploy/examples/operator-openshift.yaml:534]: hostname should be "kubernetes.io/hostname", 

lets keep 
`CSI_TOPOLOGY_DOMAIN_LABELS: "kubernetes.io/hostname, topology.kubernetes.io/region, topology.kubernetes.io/zone, topology.rook.io/rack"`

Your opinion @travisn ?

File: deploy/examples/operator.yaml [hunk 10]
```diff
@@ -449,6 +449,14 @@ data:
...
   # The CSI GRPC timeout value (in seconds). It should be >= 120. If this variable is not set or is an invalid value, it's default to 150.
   CSI_GRPC_TIMEOUT_SECONDS: "150"
+
+  # Enable topology based provisioning.
+  CSI_ENABLE_TOPOLOGY: "false"
+  # Domain labels define which node labels to use as domains
+  # for CSI nodeplugins to advertise their domains
+  # NOTE: the value here serves as an example and needs to be
+  # updated with node labels that define domains of interest
+  # CSI_TOPOLOGY_DOMAIN_LABELS: "topology.kubernetes.io/zone,topology.kubernetes.io/hostname,topology.rook.io/rack"
```
Comment (malayparida2000) [inline deploy/examples/operator.yaml:459]: Same as above 
hostname should be "kubernetes.io/hostname",

lets keep
`CSI_TOPOLOGY_DOMAIN_LABELS: "kubernetes.io/hostname, topology.kubernetes.io/region, topology.kubernetes.io/zone, topology.rook.io/rack"`

**Review by malayparida2000 — CHANGES_REQUESTED** [review]

**Comment by Rakshith-R** [comment]
changed labels, we can link it to csi topology provisioniong doc section later @malayparida2000

**Review by malayparida2000 — APPROVED** [review]

**Comment by malayparida2000** [comment]
> changed labels, we can link it to csi topology provisioniong doc section later @malayparida2000

Looks good for now, Thanks

**Review by travisn — APPROVED** [review]

**Pull request merged**